core/src/allocation.jl

```diff
@@ -0,0 +1,503 @@
+"""Whether the given node node is flow constraining by having a maximum flow rate."""
+is_flow_constraining(node::AbstractParameterNode) = hasfield(typeof(node), :max_flow_rate)
+
+"""Whether the given node is flow direction constraining (only in direction of edges)."""
+is_flow_direction_constraining(node::AbstractParameterNode) =
+    (nameof(typeof(node)) ∈ [:Pump, :Outlet, :TabulatedRatingCurve])
+
+"""Find out whether a path exists between a start node and end node in the given graph."""
+function path_exists(graph::DiGraph, start_node_id::Int, end_node_id::Int)::Bool
+    node_ids_visited = Set{Int}()
+    stack = [start_node_id]
+
+    while !isempty(stack)
+        current_node_id = pop!(stack)
+        if current_node_id == end_node_id
+            return true
+        end
+        if !(current_node_id in node_ids_visited)
+            push!(node_ids_visited, current_node_id)
+            for outneighbor_node_id in outneighbors(graph, current_node_id)
+                push!(stack, outneighbor_node_id)
+            end
+        end
+    end
+    return false
+end
+
+"""
+Construct the JuMP.jl model for allocation.
+
+Definitions
+-----------
+- 'subnetwork' is used to refer to the original Ribasim subnetwork;
+- 'AG' is used to refer to the allocation graph.
+
+Inputs
+------
+p: Ribasim model parameters
+subnetwork_node_ids: the model node IDs that are part of the allocation subnetwork
+source_edge_ids:: The IDs of the edges in the subnetwork whose flow fill be taken as
+    a source in allocation
+Δt_allocation: The timestep between successive allocation solves
+
+Outputs
+-------
+An AllocationModel object, see solve.jl.
+
+"""
+function get_allocation_model(
```

Comment:
You could also rename this to `AllocationModel` to signify it is a constructor function, similar to the ones in `create.jl`.